Sources/SwiftBSON/ExtendedJSONDecoder.swift

```diff
@@ -13,6 +15,32 @@ public class ExtendedJSONDecoder {
         return formatter
     }()
 
+    /// A set of all the possible extenededJSON wrapper keys.
+    private static var wrapperKeySet: Set<String> = {
+        Set(ExtendedJSONDecoder.wrapperKeyMap.keys)
+    }()
+
+    /// A map from extended JSON wrapper keys (e.g. "$numberLong") to the BSON type(s) that they correspond to.
+    ///
+    /// Some types are associated with multiple wrapper keys (e.g. "$code" and "$scope" both map to
+    /// `BSONCodeWithScope`). Some wrapper keys are associated with multiple types (e.g. "$code" maps to both
+    /// `BSONCode` and `BSONCodeWithScope`). Attempt to decode each of the types returned from the map until one works
+    /// to find the proper decoding.
+    private static var wrapperKeyMap: [String: [BSONValue.Type]] = {
+        var map: [String: [BSONValue.Type]] = [:]
+        for t in BSON.allBSONTypes.values {
+            for k in t.extJSONTypeWrapperKeys {
+                if var existingList = map[k] {
```

Comment:
this is a nit, but since we're focused on perf anyway here... I think this _might_ be a bit more efficient since it won't incur CoW behavior for the array(I think that would happen as written since you modify `existingList` before writing it back):
```swift
map[k, default: []].append(t.self)
```

I always thought this syntax wouldn't properly append the element in place, but I learned from doing the advent of code puzzles in Swift that it does  🤷‍♀️